benchmark/dataset_api.py

```diff
@@ -0,0 +1,337 @@
+    Tags representing properties / capabilities of this dataset.
+    """
+
+    @abstractmethod
+    def create_dataset(self, cache_dir: Path) -> Dataset:
+        """
+        Create the dataset.
+
+        This method should be deterministic - if called multiple times it should return the same
+        data. If you need randomness, please use a fixed seed.
+
+        :param cache_dir: A directory you can use to read/write any cache data.
+        """
+
+
+DATASET_FACTORIES: TaggedRegistry[DatasetFactory, DatasetTag] = TaggedRegistry()
+
+
+DatasetFactoryFn = Callable[[Path], Dataset]
+"""
+A function that can be used as a :class:`ModelFactory`.
+
+Takes a cache directory, and produces a :class:`Dataset`.
+"""
+
+
+class FnDatasetFactory(DatasetFactory):
+    """
+    Adapter from a function to a :class:`DatasetFactory`.
+    """
+
+    def __init__(self, name: str, tags: AbstractSet[DatasetTag], fn: DatasetFactoryFn) -> None:
+        self.name = name
+        self.tags = tags
+        self._fn = fn
+
+    def create_dataset(self, cache_dir: Path) -> Dataset:
+        return self._fn(cache_dir)
+
+
+def dataset_factory(
+    tags: AbstractSet[DatasetTag],
+) -> Callable[[DatasetFactoryFn], FnDatasetFactory]:
+    """
+    Decorator for turning a function into a :class:`DatasetFactory`.
+    """
+
+    def wrap(fn: DatasetFactoryFn) -> FnDatasetFactory:
+        name = fn.__name__
+        factory = FnDatasetFactory(name, tags, fn)
+        DATASET_FACTORIES.add(factory)
+        return factory
+
+    return wrap
+
+
+@check_shapes(
+    "X: [N, D]",
+    "return: [N, D]",
+)
+def _normalise(X: AnyNDArray) -> AnyNDArray:
+    X_mean = np.average(X, axis=0)[None, :]
+    X_std = 1e-6 + np.std(X, axis=0)[None, :]
+    return (X - X_mean) / X_std  # type: ignore[no-any-return]
+
+
+@check_shapes(
+    "X: [N, D]",
+    "Y: [N, P]",
+)
+def _create_dataset(
+    name: str,
+    tags: AbstractSet[DatasetTag],
+    test_fraction: float,
+    normalise: bool,
+    X: AnyNDArray,
+    Y: AnyNDArray,
+) -> Dataset:
+    """
+    Creates a :class:`Dataset` from raw data.
+    """
+    if normalise:
+        X = _normalise(X)
+        Y = _normalise(Y)
+
+    N = X.shape[0]
+    rng = np.random.default_rng(20220722)
+    indices = rng.permutation(N)
+    N_test = round(test_fraction * N)
+    train_indices = indices[N_test:]
+    test_indices = indices[:N_test]
+    train = XYData(tags, X=X[train_indices], Y=Y[train_indices])
+    test = XYData(tags, X=X[test_indices], Y=Y[test_indices])
+    return Dataset(name=name, tags=tags, train=train, test=test)
+
+
+def _get_data_file_from_url(cache_dir: Path, url: str, archive_member: Optional[Path]) -> Path:
+    file_name = url.rsplit("/", 1)[-1]
+    file_path = cache_dir / file_name
+    if not file_path.is_file():
+        cache_dir.mkdir(parents=True, exist_ok=True)
+        with urlopen(url) as response:
+            data = response.read()
+        file_path.write_bytes(data)
+        if archive_member:
```

Comment:
Not 100% sure what archive_member is doing, but why not just extract the file you're returning?